src/utils/getPartialCheckFromImportDiffs.ts

```diff
@@ -0,0 +1,89 @@
+import { reportWarning } from '../core/result';
+import NormalizedPath from '../types/NormalizedPath';
+import { FenceAndImportDiffs } from './getFenceAndImportDiffsFromGit';
+import * as path from 'path';
+import { PartialCheck } from '../types/PartialCheck';
+
+export function getPartialCheckFromImportDiffs(
+    graphDiff: FenceAndImportDiffs
+): PartialCheck | null {
+    let fences = new Set<NormalizedPath>();
+    let sourceFiles = new Set<NormalizedPath>();
+
+    let canResolve = true;
+
+    for (let [normalizedSourceFilePath, importDiff] of graphDiff.sourceImportDiffs.entries()) {
+        if (importDiff.addedImports?.length) {
+            // we need to re-check this file, since the new imports
+            // might violate the importing fence.
+            sourceFiles.add(normalizedSourceFilePath);
+        }
+    }
+
+    for (let [normalizedFencePath, fenceDiff] of graphDiff.fenceDiffs.entries()) {
+        if (fenceDiff.exports?.removed?.length) {
+            // if we removed an export, we have to re-evaluate all importers
+            // which mean we can't resolve from the repo diff
+            reportWarning(
+                `Cannot perform partial evaluation -- removed export(s) ${fenceDiff.exports.removed
+                    .map(x => {
+                        const v = { ...x };
+                        if (v.accessibleTo === null) {
+                            delete v.accessibleTo;
+                        }
+                        return JSON.stringify(v);
+                    })
+                    .join(', ')} from fence ${path.relative(process.cwd(), normalizedFencePath)}`
+            );
+            canResolve = false;
+        }
+
+        const fenceHadExportsSectionAdded =

```

Comment:
Maybe I'm missing something, but it doesn't seem like this distinguishes between:

`exports: undefined` -> `exports: ['src/index']`

and

`exports: ['src/index']` -> `exports: ['src/index', 'src/foo']`

The first case makes this _more_ restrictive, which we can't validate from the diff.  The latter makes it _less_ restrictive, which should be fine.

Reply:
This is what you mentioned in your top level comment, isn't it?  But I wasn't sure if it's supposed to be fixed at this point or not.

Reply:
we could make this explicit by making these fields non-nullable on fenceDiff. e.g. type FenceDiff as

```ts
type FenceDiff = {
  exports: 'no-diff' | {
    added: 'empty-section' | string[]
    removed: 'empty-section' | string[]
  }
  // ... etc
}
```

I personally think explicitly setting the fields as null on the diff type to say "there was no change here" or "there was no section here" represents the information in a fine way. But I'm totally willing to refactor if you think the readability would be worth the minor perf tradeoff to use strings instead.

Reply:
@Adjective-Object 

Okay, I get it.  My confusion stemmed from the fact that it's not enough to look at `added` to know what's been added; the `removed` field holds some necessary additional context.  Or vice versa when removing a rule.  It took me a bit, but now that I understand how the information is getting encoded, I'm fine with it.

Could you just add some comments to that effect in `getFenceDiff`: that "removing" a `null` means the section has been added, etc.